### ISSUE TITLE
Question about the loss function

### ISSUE THREAD

**Issue by zhang-qiang-github** [endernewton/tf-faster-rcnn#248]
Thank you for sharing your code, and I have a question about the loss function.

In the _add_losses function (line 247 in lib/nets/networks.py), you define the loss function for faster rcnn.
In line 253, you define the score of classification as:
`rpn_cls_score = tf.reshape(tf.gather(rpn_cls_score, rpn_select), [-1, 2])`
But, I really can not figure out why do you gather the original rpn_cls_score and rpn_select as the new rpn_cls_score ? Could you please give some explanations?

Thank you in advance. Your explanations will help me a lot.

Best,
Qiang

**Comment by endernewton** [comment]
oh it is just that tensorflow is building a computing graph. you may want to learn more about tf.

**Comment by zhang-qiang-github** [comment]
Thank you for your generous answer.
Sorry for mis-understanding the meaning of tf.gather. I have figure out what's the meaning of tf.gather, but I still feel confused about the code `rpn_cls_score = tf.reshape(tf.gather(rpn_cls_score, rpn_select), [-1, 2])`
In you code, `rpn_select = tf.where(tf.not_equal(rpn_label, -1))`. Why do you select the samples that label is -1? What does it means when label is -1 (and 0/1)?

**Comment by endernewton** [comment]
some labels are don't-care labels, this is a hack.
